### PR TITLE
Add failing deduce tests for array from tuple and type inside a rewrite constraint

### PR DESCRIPTION
**Pull request by danakj**
Neither can deduce the implicit parameter right now, but they should be able to.

### CODE REVIEW

**Review by geoffromer — APPROVED** [review]

File: toolchain/check/testdata/deduce/where.carbon [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+//
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/none.carbon
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/deduce/where.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/deduce/where.carbon
+
+// --- where.carbon
```
Comment (geoffromer) [inline toolchain/check/testdata/deduce/where.carbon]:
```suggestion
// --- fail_todo_where.carbon
```

File: toolchain/check/testdata/deduce/where.carbon [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+interface Z {
+  let Z1:! type;
+}
+
+class C(T:! type) {}
+
+fn F[U:! type](T:! Z where .Z1 = C(U)) {}
+
+class D {}
+final impl C(()) as Z where .Z1 = C(D) {}
```
Comment (dwblaikie) [inline toolchain/check/testdata/deduce/where.carbon]: Am I reading this right that `C(())` could be any type here (doesn't have any load bearing relationship to `C(D)` or `Z`, etc, here)? Could it use some other type, then - might be a bit clearer what's related and what isn't?

**Review by dwblaikie — APPROVED** [review]

File: toolchain/check/testdata/deduce/where.carbon (hunk 2 above)
Comment (danakj) [reply toolchain/check/testdata/deduce/where.carbon]: Yes, done.

File: toolchain/check/testdata/deduce/where.carbon (hunk 1 above)
Comment (danakj) [reply toolchain/check/testdata/deduce/where.carbon]: Done
